plugins/inputs/jenkins/jenkins.go

```diff
@@ -0,0 +1,483 @@
+package jenkins
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/internal/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/kelwang/gojenkins"
+)
+
+// Jenkins plugin gathers information about the nodes and jobs running in a jenkins instance.
+type Jenkins struct {
+	URL      string
+	Username string
+	Password string
+	// HTTP Timeout specified as a string - 3s, 1m, 1h
+	ResponseTimeout internal.Duration
+
+	tls.ClientConfig
+	instance *gojenkins.Jenkins
+
+	MaxTCPConcurrentConnections int               `toml:"max_tcp_concurrent_connections"`
+	MaxBuildAge                 internal.Duration `toml:"max_build_age"`
+	MaxSubJobsLayer             int               `toml:"max_sub_jobs_layer"`
+	NewestSubJobsEachLayer      int               `toml:"newest_sub_jobs_each_layer"`
+	JobExclude                  []string          `toml:"job_exclude"`
+	jobFilter                   map[string]bool
+
+	NodeExclude []string `toml:"node_exclude"`
+	nodeFilter  map[string]bool
+}
+
+type byBuildNumber []gojenkins.JobBuild
+
+const sampleConfig = `
+url = "http://my-jenkins-instance:8080"
+username = "admin"
+password = "admin"
+## Set response_timeout
+response_timeout = "5s"
+
+## Optional SSL Config
+# ssl_ca = /path/to/cafile
+# ssl_cert = /path/to/certfile
+# ssl_key = /path/to/keyfile
+## Use SSL but skip chain & host verification
+# insecure_skip_verify = false
+
+## Job & build filter
+# max_build_age = "1h"
+## jenkins can have unlimited layer of sub jobs
+## this config will limit the layers of pull, default value 0 means
+## unlimited pulling until no more sub jobs
+# max_sub_jobs_layer = 0
+## in workflow-multibranch-plugin, each branch will be created as a sub job
+## this config will limit to call only the lasted branches
+## sub jobs fetch in each layer
+# empty will use default value 10
+# newest_sub_jobs_each_layer = 10
+# job_exclude = [ "MyJob", "MyOtherJob" ]
+
+## Node filter
+# node_exlude = [ "node1", "node2" ]
+
+## Woker pool for jenkins plugin only
+# empty this field will use default value 30
+# max_tcp_concurrent_connections = 30
+`
+
+// measurement
+const (
+	measurementNode = "jenkins_node"
+	measurementJob  = "jenkins_job"
+)
+
+type typedErr struct {
```

Comment:
typedErr isn't a very good name because it doesn't describe what type of error it is.  Try to come up with a more descriptive name or just call it Error.

Reply:
👍 changed to Error